rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,697 @@
+
+By checking out how the code is done in quite a few git implementations out there. Some have already implemented it. In the worst case, you can read main [Git code](https://github.com/git/git) and figure out how those commands work from first principles and implement them.
+
+Paying attention that we are running in a browser context and not all details are needed, simplifies things a lot.
+
+#### Q: Will using Isomorphic Git answer to all stated goals above?
+
+✅ = already done
+🚧 = work required
+❎ = not possible
+
+Goal 1: Must run in the browser/on the client [High Confidence] ✅
+
+Goal 2: Lazy loading of files and git history [High Confidence] 🚧
+
+Is achievable in a few ways. But in current state would require sparse checkout to be implemented.
+
+Goal 3: Must be git compatible but not necessarily up to spec [Medium Confidence] ✅
+
+No issue here. Isomorphic Git is already git compatible, if we decide to extend it with new features we can have a fork of Isomorphic Git potentially or bring commands up to spec and merge them into Isomorphic Git repo.
+
+Goal 4: (Future) File-based auth [High Confidence | Server-related ] 🚧
+
+Is part of server so outside of Iso Git scope.
+
+> note: maybe wrong
+
+Goal 5: (Future) Support for large files out of the box [Medium Confidence] 🚧
+
+> note: need to research more to give good estimate how doable this is to add
+
+There is [open issue on this](https://github.com/isomorphic-git/isomorphic-git/issues/1375).
+
+Goal 6: (Future) Plugin system to support different files [Medium Confidence] 🚧
+
+Would be doable to add irrespective of whether libgit2 or Isomorphic Git is chosen.
+
+#### Q: Could Git SDK abstract away the file system and expose git app focused API only?
+
+> note: can be removed from RFC, just some thinking out loud
+
+> also on second thought this won't make sense, seems you would need to pass a FS
+
+> hypothethisizing making Git SDK have tight integration with the FS (might not be possible/feasible or make sense)
+
+> only exposing the actual useful commands you would need to build Git based apps
+
+Perhaps outside of the discussion of this RFC but still maybe interesting to discuss. Or perhaps create new RFC?
+
+It would be potentially interesting to see a Git SDK that would also abstract working with the file system all together.
+
+This would mean, there would be no need to pass `fs` to every git operation. You would just do:
+
+```js
+let inlang = await gitSDK.clone("https://github.com/inlang/inlang.git")
+```
+
+And you get back a fully fledged file system fetched from the repo. As the SDK is focused on running on context of browser.
+
+Above function call can do a fetch of just enough resources to do further actions.
+
+`git clone --depth 1 --sparse --no-checkout --filter=blob:none https://github.com/inlang/inlang`
+
+After this an app would most likely want to fetch a file to edit. It can do:
+
```

Comment:
Abstracting the filesystem away makes the git sdk lose one of its appeals. 

Files (and thereby filesystems) are amazing and enable inlang to exist. The VSCode extension uses a local filesystem provided by VSCode. The browser uses a filesystem that fetches files remotely. The CLI uses a filesystem provided by node.    

If the git-sdk would abstract away the filesystem, using the git-sdk outside of the browser becomes a burden. If users have their files locally on disk, there is no appealing reason to force a remote filesystem.